components/renku_data_services/secrets/core.py

```diff
@@ -0,0 +1,134 @@
+
+    decrypted_secrets = {}
+    try:
+        for secret in secrets:
+            try:
+                decryption_key = decrypt_rsa(secret_service_private_key, secret.encrypted_key)
+            except ValueError:
+                if previous_secret_service_private_key is not None:
+                    # If we're rotating keys right now, try the old key
+                    decryption_key = decrypt_rsa(previous_secret_service_private_key, secret.encrypted_key)
+                else:
+                    raise
+
+            decrypted_value = decrypt_string(decryption_key, user.id, secret.encrypted_value).encode()  # type: ignore
+            decrypted_secrets[secret.name] = b64encode(decrypted_value).decode()
+    except Exception as e:
+        # don't wrap the error, we don't want secrets accidentally leaking.
+        raise errors.SecretDecryptionError(message=f"An error occured decrypting secrets: {str(type(e))}")
+
+    owner_refs = []
+    if owner_references:
+        owner_refs = [o.to_k8s() for o in owner_references]
+    secret = k8s_client.V1Secret(
+        data=decrypted_secrets,
+        metadata=k8s_client.V1ObjectMeta(
+            name=secret_name,
+            namespace=namespace,
+            owner_references=owner_refs,
+        ),
+    )
+
+    try:
+        core_client.create_namespaced_secret(namespace, secret)
+    except k8s_client.ApiException as e:
+        # don't wrap the error, we don't want secrets accidentally leaking.
+        raise errors.SecretCreationError(message=f"An error occured creating secrets: {str(type(e))}")
+
+    return secret_name
+
+
+async def rotate_encryption_keys(
+    new_key: rsa.RSAPrivateKey, old_key: rsa.RSAPrivateKey, secrets_repo: UserSecretsRepo, batch_size: int = 100
+) -> None:
+    """Rotate all secrets to a new private key.
+
+    This method undoes the outer encryption and reencrypts with a new key, without touching the inner encryption.
+    """
+    processed_secrets_metrics = Counter(
+        "secrets_rotation_count",
+        "Number of secrets rotated",
+    )
+    runnning_metrics = Enum(
+        "secrets_rotation_state", "State of secrets rotation", states=["running", "finished", "errored"]
+    )
+    runnning_metrics.state("running")
+    try:
+        async for batch in secrets_repo.get_all_secrets_batched(batch_size):
+            updated_secrets = []
+            for secret, user_id in batch:
+                new_secret = await rotate_single_encryption_key(secret, user_id, new_key, old_key)
+                if new_secret is not None:
+                    updated_secrets.append(new_secret)
+
+            await secrets_repo.update_secrets(updated_secrets)
+            processed_secrets_metrics.inc(len(updated_secrets))
+    except:
+        runnning_metrics.state("errored")
+        raise
+    else:
+        runnning_metrics.state("finished")
+
+
+async def rotate_single_encryption_key(
```

Comment:
well it enables cooperative multitasking. as in, every time this is awaited, even if all the code it blocking, the runtime can decide to run something else instead. Otherwise it would block completely until all secrets of a batch are done. I.e. it's us telling python "when this function gets called it's a nice opportunity to run other stuff as well".

we can also opt for using https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor

Reply:
Note that the default behavior in asyncio, as far as I understand, is to run the task that has been ready for the longest(it uses a deque internally). So if a request comes in during a call to this function, the request should be handled before this function is executed again.